sdk/digitaltwins/Azure.DigitalTwins.Core/src/Models/DigitalTwinMetadataJsonConverter.cs

```diff
@@ -23,23 +23,22 @@ public override DigitalTwinMetadata Read(ref Utf8JsonReader reader, Type typeToC
 
             if (reader.TokenType != JsonTokenType.StartObject)
             {
-                throw new JsonException($"Unexpected token type {reader.TokenType} at index {reader.TokenStartIndex} Expected JsonTokenType.StartObject.");
+                throw new JsonException($"Unexpected token type {reader.TokenType} at index {reader.TokenStartIndex}. Expected JsonTokenType.StartObject.");
             }
 
-            var metadata = new DigitalTwinMetadata();
             reader.Read(); // Advance into our object.
+            var metadata = new DigitalTwinMetadata();
 
             // Until we reach the end of the object we began reading
             while (reader.TokenType != JsonTokenType.EndObject)
             {
-                string tokenName = reader.GetString();
-
                 if (reader.TokenType != JsonTokenType.PropertyName)
                 {
-                    throw new JsonException($"Unexpected token {reader.TokenType} {tokenName} found in DigitalTwinMetadata at index {reader.TokenStartIndex}, position {reader.Position}");
+                    throw new JsonException();
```

Comment:
I believe, for any valid JSON input, it is guaranteed `reader.TokenType` will be a propertyname here, and hence this check can be removed.

For invalid JSON input (that doesn't follow proper JSON syntax), `reader.Read()` will throw already.

The loop condition you have is "keep reading the JSON until we hit end object". Before the loop, you validate that the token is a start object. Valid JSON objects are a list of `property name : value` pairs.

So, when you call `reader.Read()`, while inside a JSON object, the first token has to be a property name (assuming default options settings on the serializer so that comments are considered invalid).

Then, in your loop body, you grab the property name, move to the next token, and expect either a string (when property name is $model) or another JSON object which you read through fully as part of the recursive JsonSerializer.Deserialize call.

The only valid token after the next `reader.Read()` call at the end of the loop has to either be another property name or end object.

**Unless you have a test case or scenario which showcases otherwise, this if-check seems unreachable to me and can be removed.**

Reply:
Thanks for that insight. I've removed that check.